File: src/BesselK.hs
module BesselK (
  besselK1x
) where

import Data.Array

-- Following https://arxiv.org/abs/1209.1547
besselK :: Double -> Double -> Double
besselK ν 0 = 1/0
besselK ν z = tryH 1 (withH 2)
    where tryH h prev = let new = withH h in if abs (new - prev) <= 1e-11 * prev then new else tryH (h/2) new
          withH h = (*h) $ sum $ (exp (-z) / 2 :) $ takeWhile (> exp (-z) * 1e-20) $ map (\n -> cosh (ν*n*h) * exp (-z * cosh (n*h))) [1..]

data Quartic = Quartic Double Double Double Double Double

evalQuartic :: Quartic -> Double -> Double
evalQuartic (Quartic a b c d e) x = a + x*(b + x*(c + x*(d + x*e)))

-- taylor series of x^2 K_1(x^2)
besselK1x2TaylorSeries :: Double -> Quartic
besselK1x2TaylorSeries x = Quartic
    (x ^ 2 * k1)
    (-2 * x^3 * k0)
    ((4 * x^4 * k1 - 6 * x^2 * k0)/2)
    ((20 * x^3 * k1 - (8 * x^5 + 12 * x) * k0)/6)
    (((16 * x^6 + 44*x^2) * k1 - (80 * x^4 + 12) * k0)/24)
  where k0 = besselK 0 (x ^ 2)
        k1 = besselK 1 (x ^ 2)

gridFineness = 1e-3
gridMax = 6

-- I thought using an unboxed array of doubles might be quicker here, but when I benchmarked it in GHCI, the array of `Quartic`s turned out to actually be a little faster.
memoized :: Array Int Quartic
memoized = listArray (0, limit) [besselK1x2TaylorSeries ((fromIntegral x + 0.5)*gridFineness) | x <- [0..limit]]
  where limit = ceiling (gridMax/gridFineness) - 1

-- x K1(x), using the memoized table of taylor expansions.
besselK1x :: Double -> Double
besselK1x x | x >= gridMax * gridMax = 0
besselK1x x | otherwise    = evalQuartic (memoized ! i) dx
  where i = round (sqrt x/gridFineness - 0.5)
        dx = sqrt x - (fromIntegral i + 0.5) * gridFineness

File: Rust/src/bessel_k.rs
use std::array;
use std::sync::LazyLock;

// Following https://arxiv.org/abs/1209.1547
fn bessel_K(ν: f64, z: f64) -> f64 {
  #![allow(non_snake_case)]
  if z == 0.0 {
    return f64::INFINITY;
  }
  let mut prev_result : f64 = -2.0;
  let mut result = -1.0;
  let mut h = 1.0;
  while (result - prev_result).abs() > 1e-11 * prev_result.abs() {
    h /= 2.0;
    prev_result = result;
    result = 0.0;
    let start = (1..).filter(|n| (ν*(*n as f64)*h).cosh() * (-z * ((*n as f64)*h).cosh()).exp() < (-z).exp() * 1e-20).next().unwrap();
    for n in (1..start).rev() {
      result += (ν*(n as f64)*h).cosh() * (-z * ((n as f64)*h).cosh()).exp();
    }
    result += (-z).exp() / 2.0;
    result *= h;
  }
  result
}

struct Quartic{a:f64, b:f64, c:f64, d:f64, e:f64}

impl Quartic {
  fn eval(&self, x:f64) -> f64 {
    self.a + x*(self.b + x*(self.c + x*(self.d + x*self.e)))
  }
}

// taylor series of x^2 K_1(x^2)
fn bessel_K1_x2_taylor_series(x: f64) -> Quartic {
  #![allow(non_snake_case)]
  let k0 = bessel_K(0.0, x*x);
  let k1 = bessel_K(1.0, x*x);
  Quartic {
    a: x*x * k1,
    b: -2.0 * x.powi(3) * k0,
    c: (4.0 * x.powi(4) * k1 - 6.0 * x*x * k0)/2.0,
    d: (20.0 * x.powi(3) * k1 - (8.0 * x.powi(5) + 12.0 * x) * k0)/6.0,
    e: ((16.0 * x.powi(6) + 44.0*x*x) * k1 - (80.0 * x.powi(4) + 12.0) * k0)/24.0,
  }
}

const GRID_FINENESS : f64 = 1e-3;
const GRID_MAX : f64 = 6.0;
const GRID_POINTS : usize = (GRID_MAX/GRID_FINENESS).ceil() as usize;

static SERIES_TABLE : LazyLock<[Quartic; GRID_POINTS]> = LazyLock::new(|| array::from_fn(|n| bessel_K1_x2_taylor_series((n as f64 + 0.5)*GRID_FINENESS)));

// x K1(x), using the memoized table of taylor expansions.
pub fn bessel_K1_x(x: f64) -> f64 {
  #![allow(non_snake_case)]
  if x >= GRID_MAX * GRID_MAX {
    return 0.0;
  } else if x < 0.0 {
    panic!("Bessel K cannot be evaluated for negative arguments.");
  }
  let i = (x.sqrt()/GRID_FINENESS - 0.5).round();
  let dx = x.sqrt() - (i + 0.5) * GRID_FINENESS;
  (*SERIES_TABLE)[i as usize].eval(dx)
}
